src/transformers/models/relation_detr/modeling_relation_detr.py

```diff
@@ -0,0 +1,2788 @@
+    """
+
+    def __init__(self, n):
+        super().__init__()
+        self.register_buffer("weight", torch.ones(n))
+        self.register_buffer("bias", torch.zeros(n))
+        self.register_buffer("running_mean", torch.zeros(n))
+        self.register_buffer("running_var", torch.ones(n))
+
+    def _load_from_state_dict(
+        self, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+    ):
+        num_batches_tracked_key = prefix + "num_batches_tracked"
+        if num_batches_tracked_key in state_dict:
+            del state_dict[num_batches_tracked_key]
+
+        super()._load_from_state_dict(
+            state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+        )
+
+    def forward(self, x):
+        # move reshapes to the beginning
+        # to make it user-friendly
+        weight = self.weight.reshape(1, -1, 1, 1)
+        bias = self.bias.reshape(1, -1, 1, 1)
+        running_var = self.running_var.reshape(1, -1, 1, 1)
+        running_mean = self.running_mean.reshape(1, -1, 1, 1)
+        epsilon = 1e-5
+        scale = weight * (running_var + epsilon).rsqrt()
+        bias = bias - running_mean * scale
+        return x * scale + bias
+
+
+# Copied from transformers.models.detr.modeling_detr.replace_batch_norm with Detr->RelationDetr
+def replace_batch_norm(model):
+    r"""
+    Recursively replace all `torch.nn.BatchNorm2d` with `RelationDetrFrozenBatchNorm2d`.
+
+    Args:
+        model (torch.nn.Module):
+            input model
+    """
+    for name, module in model.named_children():
+        if isinstance(module, nn.BatchNorm2d):
+            new_module = RelationDetrFrozenBatchNorm2d(module.num_features)
+
+            if not module.weight.device == torch.device("meta"):
+                new_module.weight.data.copy_(module.weight)
+                new_module.bias.data.copy_(module.bias)
+                new_module.running_mean.data.copy_(module.running_mean)
+                new_module.running_var.data.copy_(module.running_var)
+
+            model._modules[name] = new_module
+
+        if len(list(module.children())) > 0:
+            replace_batch_norm(module)
+
+
+class RelationDetrConvEncoderPostProcess(nn.Module):
+    """Post process for the ConvEncoder. This is used to be compatible with the FocalNet backbone in official repo."""
+
+    def __init__(self, in_channels: int, post_layer_norm: bool = False):
+        super().__init__()
+        self.in_channels = in_channels
+        self.post_layer_norm = post_layer_norm
+        if self.post_layer_norm:
+            for idx, channel in enumerate(in_channels):
+                self.add_module(f"norm{idx}", nn.LayerNorm(channel))
+
+    def forward(self, multi_level_feats: List[Tensor]):
+        if self.post_layer_norm:
+            # convert N C H W to N H W C
+            if all(feat.shape[1] == channel for feat, channel in zip(multi_level_feats, self.in_channels)):
```

Comment:
Why do wee need `if` here?do we have both cases for the provided model?

Reply:
This is for the FocalNet backbone. It gives outputs in NCHW format, we need to convert them to NHWC for layer norm, and them convert them back to NCHW format as output. To be clear:

```python
resnet (NCHW) -> output
swin   (NHWC) -> (NHWC to NCHW) -> output
focal  (NCHW) -> (NCHW to NHWC) -> layer_norm -> (NHWC to NCHW) -> output
```